presto-main/src/main/java/com/facebook/presto/operator/TableWriterOperator.java

```diff
@@ -151,40 +170,45 @@ public OperatorContext getOperatorContext()
     @Override
     public void finish()
     {
+        ListenableFuture<?> currentlyBlocked = blocked;
+        statisticAggregation.finish();
+        ListenableFuture<?> blockedOnAggregation = statisticAggregation.isBlocked();
+        ListenableFuture<?> blockedOnFinish = NOT_BLOCKED;
         if (state == State.RUNNING) {
             state = State.FINISHING;
             finishFuture = pageSink.finish();
-            blocked = toListenableFuture(finishFuture);
+            blockedOnFinish = toListenableFuture(finishFuture);
             updateWrittenBytes();
         }
+        if (currentlyBlocked.isDone() && blockedOnAggregation.isDone() && blockedOnFinish.isDone()) {
+            this.blocked = NOT_BLOCKED;
```

Comment:
nit: unnecessary optimization, `else`'s code would cover that

Reply:
I didn't see any shortpath for the all finished case? Is there one?

Reply:
Actually it doesn't have a shortpath, but the "long" path for a completed future is not bad. It just has to create and execute the callbacks in place. Let me remove this optimization.

Reply:
Yes, the shortpath isn't that short, but this is going to be executed infrequently.